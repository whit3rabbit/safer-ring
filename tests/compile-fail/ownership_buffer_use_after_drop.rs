//! This test verifies that OwnedBuffer cannot be used after being moved
//! into an operation, ensuring the ownership transfer model works correctly.

use safer_ring::{Ring, OwnedBuffer};

#[tokio::main]
async fn main() {
    let mut ring = Ring::new(32).unwrap();
    let buffer = OwnedBuffer::new(1024);
    
    // Move buffer into operation
    let _future = ring.read_owned(0, buffer);
    
    // This should fail to compile - buffer has been moved
    println!("Buffer size: {}", buffer.size()); //~ ERROR
}