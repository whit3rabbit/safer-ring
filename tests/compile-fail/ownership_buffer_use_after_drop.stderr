warning: variable does not need to be mutable
 --> tests/compile-fail/ownership_buffer_use_after_drop.rs:8:9
  |
8 |     let mut ring = Ring::new(32).unwrap();
  |         ----^^^^
  |         |
  |         help: remove this `mut`
  |
  = note: `#[warn(unused_mut)]` on by default

error[E0382]: borrow of moved value: `buffer`
  --> tests/compile-fail/ownership_buffer_use_after_drop.rs:15:33
   |
9  |     let buffer = OwnedBuffer::new(1024);
   |         ------ move occurs because `buffer` has type `OwnedBuffer`, which does not implement the `Copy` trait
...
12 |     let _future = ring.read_owned(0, buffer);
   |                                      ------ value moved here
...
15 |     println!("Buffer size: {}", buffer.size()); //~ ERROR
   |                                 ^^^^^^ value borrowed here after move
